### PR TITLE
Upload toolstates.json to rust-lang-ci2

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR does two things:

* Following up with https://github.com/rust-lang/rust/pull/65202, it migrates deploying artifacts to CI in a script. Both uploading release artifacts and CPU stats were merged into the same script, designing it to be easily extended.
* Uploads the toolstate JSON to `rust-lang-ci2` along with the release artifacts, both for Linux and Windows. This is needed because @RalfJung wants to stop shipping MIRI when its tests are failing, and the toolstate repo doesn't have entries for each commit. Having the toolstate data (just for that specific commit) on `rust-lang-ci2` will simplify the code a lot.

r? @alexcrichton 
cc @RalfJung

### CODE REVIEW

**Comment by pietroalbini** [comment]
@alexcrichton did you also review 25e606978ab6417fa93cc19099e05f16b3081d83?

**Comment by pietroalbini** [comment]
Rebased this on top of master.

@bors r=alexcrichton

File: src/ci/scripts/upload-artifacts.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+if [[ ! -z "${DEPLOY_TOOLSTATES_JSON+x}" ]]; then
+    cp /tmp/toolstates.json "${upload_dir}/${DEPLOY_TOOLSTATES_JSON}"
+fi
+
+echo "Files that will be uploaded:"
+ls -lah "${upload_dir}"
+echo
+
+deploy_dir="rustc-builds"
+if [[ "${DEPLOY_ALT-0}" = "1" ]]; then
```
Comment (lzutao) [inline src/ci/scripts/upload-artifacts.sh]:
```suggestion
if [[ "${DEPLOY_ALT-0}" -eq "1" ]]; then
```

File: src/ci/scripts/upload-artifacts.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    fi
+    rm -rf "${dist_dir}/doc"
+    cp -r "${dist_dir}"/* "${upload_dir}"
+fi
+
+# CPU usage statistics.
+cp cpu-usage.csv "${upload_dir}/cpu-${CI_JOB_NAME}.csv"
+
+# Toolstate data.
+if [[ ! -z "${DEPLOY_TOOLSTATES_JSON+x}" ]]; then
```
Comment (lzutao) [inline src/ci/scripts/upload-artifacts.sh]:
```suggestion
if [[ -n "${DEPLOY_TOOLSTATES_JSON+x}" ]]; then
```

File: src/ci/scripts/upload-artifacts.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+set -euo pipefail
+IFS=$'\n\t'
+
+source "$(cd "$(dirname "$0")" && pwd)/../shared.sh"
+
+upload_dir="$(mktemp -d)"
+
+# Release tarballs produced by a dist builder.
+if [[ "${DEPLOY-0}" = "1" ]] || [[ "${DEPLOY_ALT-0}" = "1" ]]; then
```
Comment (lzutao) [inline src/ci/scripts/upload-artifacts.sh]:
```suggestion
if [[ "${DEPLOY-0}" -eq "1" ]] || [[ "${DEPLOY_ALT-0}" -eq "1" ]]; then
```

**Comment by pietroalbini** [comment]
Ok the last try build seems to be going through.

@alexcrichton I had to add https://github.com/rust-lang/rust/pull/65274/commits/ca3468768d234af5de4fe7b578b701e3404a6dac as well, could you review it and `r+ rollup` the PR once the try build finishes? The rollup is extra safe as I tested all the builders affected by this with try builds.

**Comment by alexcrichton** [comment]
@bors: r+ 

sure!

I'll leave the `rollup` bit to those who come along and see success

**Comment by RalfJung** [comment]
Thanks a lot @pietroalbini! I will get back to https://github.com/rust-lang/rust/issues/60301 based on top of this on the week-end.

**Pull request merged**